src/jsep.js

```diff
@@ -2,257 +2,524 @@
+	 * @param {string} op_name The name of the unary op to add
+	 * @return this
+	 */
+	addUnaryOp(op_name) {
+		this.max_unop_len = Math.max(op_name.length, this.max_unop_len);
+		this.unary_ops[op_name] = 1;
+		return this;
+	}
+
+	/**
+	 * @method jsep.addBinaryOp
+	 * @param {string} op_name The name of the binary op to add
+	 * @param {number} precedence The precedence of the binary op (can be a float)
+	 * @return this
+	 */
+	addBinaryOp(op_name, precedence) {
+		this.max_binop_len = Math.max(op_name.length, this.max_binop_len);
+		this.binary_ops[op_name] = precedence;
+		return this;
+	}
+
+	/**
+	 * @method addIdentifierChar
+	 * @param {string} char The additional character to treat as a valid part of an identifier
+	 * @return this
+	 */
+	addIdentifierChar(char) {
+		this.additional_identifier_chars.add(char);
+		return this;
+	}
+
+	/**
+	 * @method addLiteral
+	 * @param {string} literal_name The name of the literal to add
+	 * @param {*} literal_value The value of the literal
+	 * @return this
+	 */
+	addLiteral(literal_name, literal_value) {
+		this.literals[literal_name] = literal_value;
+		return this;
+	}
+
+	/**
+	 * @method removeUnaryOp
+	 * @param {string} op_name The name of the unary op to remove
+	 * @return this
+	 */
+	removeUnaryOp(op_name) {
+		delete this.unary_ops[op_name];
+		if (op_name.length === this.max_unop_len) {
+			this.max_unop_len = JsepConfig.getMaxKeyLen(this.unary_ops);
+		}
+		return this;
+	}
+
+	/**
+	 * @method removeAllUnaryOps
+	 * @return this
+	 */
+	removeAllUnaryOps() {
+		this.unary_ops = {};
+		this.max_unop_len = 0;
+
+		return this;
+	}
+
+	/**
+	 * @method removeIdentifierChar
+	 * @param {string} char The additional character to stop treating as a valid part of an identifier
+	 * @return this
+	 */
+	removeIdentifierChar(char) {
+		this.additional_identifier_chars.delete(char);
+		return this;
+	}
+
+	/**
+	 * @method removeBinaryOp
+	 * @param {string} op_name The name of the binary op to remove
+	 * @return this
+	 */
+	removeBinaryOp(op_name) {
+		delete this.binary_ops[op_name];
+
+		if (op_name.length === this.max_binop_len) {
+			this.max_binop_len = JsepConfig.getMaxKeyLen(this.binary_ops);
+		}
+
+		return this;
+	}
+
+	/**
+	 * @method removeAllBinaryOps
+	 * @return this
+	 */
+	removeAllBinaryOps() {
+		this.binary_ops = {};
+		this.max_binop_len = 0;
+
+		return this;
+	}
+
+	/**
+	 * @method removeLiteral
+	 * @param {string} literal_name The name of the literal to remove
+	 * @return this
+	 */
+	removeLiteral(literal_name) {
+		delete this.literals[literal_name];
+		return this;
+	}
+
+	/**
+	 * @method removeAllLiterals
+	 * @return this
+	 */
+	removeAllLiterals() {
+		this.literals = {};
+
+		return this;
+	}
+}
+
+
+// JavaScript Expression Parser
+export class Jsep {
+	static get version() {
+		// To be filled in by the template
+		return '<%= version %>';
+	}
+
+	/**
+	 * @returns {(pos: number) => string}
+	 */
+	get charAtFunc() {
+		return this.expr.charAt;
```

Comment:
What's the point of this? (in the original code as well). Why not just directly use `this.expr.charAt()` where needed?
Same for the one below.

Reply:
TBH, I wasn't sure why the original code had added those utility functions (back in 2013), but currently at:
```
let charAtFunc = expr.charAt;
let charCodeAtFunc = expr.charCodeAt;
let exprI = function(i) {
	return charAtFunc.call(expr, i);
};
let exprICode = function(i) {
	return charCodeAtFunc.call(expr, i);
};
let length = expr.length;
```
After benchmarking the getter() as significantly slower (below), I think a standard function call would be better. The code does use a lot of getting either the char or charCodeAt ... should we change all those expressions to directly use `this.expr.charCodeAt(this.index)` or `this.expr.charAt(this.index)`? Either way it's a pretty common use-case, even for plugins if we want to expose something?

Reply:
What about a getter that returns the actual **result** of `this.expr.charAt(this.index)` (and `charCodeAt`)?
We could call it `currentChar` and `currentCharCode` and I think it would make the code more readable too compared to this `exprI` stuff. What do you think?

Reply:
I agree -- except it appears that getters are 60% slower than functions, so I think just a standard method?

Reply:
I know you've observed a 60% slowdown in a particular case, but can we generalize it that much? It's a bit drastic to say let's never use getters because they're 60% slower (also, you already use getters in many places!).